### PR TITLE
Add stack persistent volume claim

### LINKED ISSUE #42: Hook artifacts not available for multi node clusters
Currently the Hook artifacts are downloaded to a `hostPath`. This is fine for single node Kubernetes clusters, but not for clusters with more than one.
 
https://github.com/tinkerbell/charts/blob/2a4408f72a44b4a2668dd33ebb4e40c1c52e3f3a/tinkerbell/stack/templates/nginx.yaml#L88

## Expected Behaviour

The stack chart should handle making the Hooks artifacts available for multiple nodes. Maybe a pluggable CSI option? Maybe, https://github.com/openebs/openebs or https://github.com/longhorn/longhorn

## Steps to Reproduce (for bugs)

1.
2.
3.
4.

## Your Environment

* Operating System and version (e.g. Linux, Windows, MacOS):

* How are you running Tinkerbell? Using Vagrant & VirtualBox, Vagrant & Libvirt, on Packet using Terraform, or give details:

* Link to your project or a code example to reproduce issue:

### PR DESCRIPTION
**Pull request by pmorillon**
## Description

A proposal to fix #42 like PR #63, taking into account the new method to download Hook artifacts with a Job (previously with an init container). PVC accessModes are checked to disable hook downloader Job and inform user when ReadWriteOnce is used (default to ReadWriteMany).

## Why is this needed

Fixes: #42 

## How Has This Been Tested?

Only tested with ReadWriteOnce PVC.
Checked generated manifests with helm install --debug --dry-run...

## How are existing users impacted? What migration steps/scripts do we need?

Actual default behavior should be preserved

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

File: tinkerbell/stack/templates/NOTES.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{{- if eq (include "stack.hookDownloadJobEnabled" .) "false" }}
+Warning:
+
+You are using no persistence or ReadWriteOnce access modes on PersistentVolumeClaim, Job `download-hook` is disabled.
+You will need to manage hooks artifacts manually.
+Also stack Deployment is not scalable.
```
Comment (jacobweinstock) [inline tinkerbell/stack/templates/NOTES.txt:6]: Hey @pmorillon, thanks for this PR. What does this mean, the stack deployment is not scalable? is this only when the HookOS download job is false? or in general?

File: tinkerbell/stack/templates/hook.yaml [hunk 2]
```diff
@@ -1,4 +1,5 @@
-{{- if .Values.stack.hook.enabled }}
+{{- $hookDownloadJobEnabled := eq (include "stack.hookDownloadJobEnabled" .) "true" -}}
+{{- if and .Values.stack.hook.enabled $hookDownloadJobEnabled }}
```
Comment (jacobweinstock) [inline tinkerbell/stack/templates/hook.yaml:2]: Its hard to tell where the corresponding `{{- end }}` is for this. mind putting the corresponding`{{- end }}` at the same level indent?

File: tinkerbell/stack/values.yaml [hunk 3]
```diff
@@ -12,6 +12,20 @@ stack:
   lbClass: kube-vip.io/kube-vip-class
   # Once the Kubernetes Gateway API is more stable, we will use that for all services instead of nginx.
   image: nginx:1.25.1
+  persistence:
```
Comment (jacobweinstock) [inline tinkerbell/stack/values.yaml:15]: if i understand correctly, persistence is only related to the HookOS download? if that is the case, would this block be more ideal under `hook:`?

File: tinkerbell/stack/templates/NOTES.txt (hunk 1 above)
Comment (pmorillon) [reply tinkerbell/stack/templates/NOTES.txt:6]: I will remove this confusing note and add comment into values file and update _helper to apply this behavior : 
* stack.hook.enabled to false : no download job, no PVC, no hook-artifacts Volume
* stack.hook.enabled to true (default) : 
  * stack.hook.persistence.enabled to false : volume is an empty dir, so no download job, no PVC
  * stack.hook.persistence.enabled to true (default) : 
    * stack.hook.persistence.type = hostpath (default) : volume is an hostpath, download job enabled, no PVC. Only works with a one worker node cluster for hook downloader and scale up.
    * stack.hook.persistence.type = pvc : volume is based on a PVC
      * stack.hook.persistent.accessModes contains ReadWriteMany (default) : download job enabled. Will works on multinode cluster and scale up is possible.
      * stack.hook.persistent.accessModes contains ReadWriteOnce : download job is disabled (job won't be able to mount already mounted volume), scale up will not be possible.

File: tinkerbell/stack/templates/hook.yaml (hunk 2 above)
Comment (pmorillon) [reply tinkerbell/stack/templates/hook.yaml:2]: All right

**Comment by jacobweinstock** [comment]
Hey @pmorillon , thanks for this. Looks good. Just going to run a few test scenarios locally and then we'll merge it.

File: tinkerbell/stack/templates/_helpers.tpl
Comment (chrisdoherty4) [inline tinkerbell/stack/templates/_helpers.tpl:1]: Can we rename this file `_hook.tpl`.

File: tinkerbell/stack/templates/hook.yaml [hunk 4]
```diff
@@ -48,9 +49,20 @@ spec:
...
+          {{- if .Values.stack.hook.persistence.enabled }}
+            {{- if eq .Values.stack.hook.persistence.type "hostpath" }}
           hostPath:
             path: {{ .Values.stack.hook.downloadsDest }}
             type: DirectoryOrCreate
+            {{- else if eq .Values.stack.hook.persistence.type "pvc" }}
+          persistentVolumeClaim:
+            claimName: {{ .Values.stack.hook.persistence.existingClaim | default "hook-artifacts" }}
+            {{- else }}
+              {{- fail "value for .Values.stack.hook.persistence.type is not as expected" }}
```
Comment (chrisdoherty4) [inline tinkerbell/stack/templates/hook.yaml:61]:
```suggestion
              {{- fail "value for .Values.stack.hook.persistence.type is unsupported" }}
```

File: tinkerbell/stack/values.yaml [hunk 5]
```diff
@@ -21,6 +23,24 @@ stack:
     # downloadURL only works with the > 0.8.1 Hook release because
     # previous Hook versions didn't provide a checksum file.
     downloadURL: https://github.com/tinkerbell/hook/releases/download/latest
+    persistence:
+      enabled: true
+      ## type hostpath or pvc
+      # hostpath : only works on a single worker node cluster
+      type: hostpath
+      ## When type is pvc
+      # storageClassName: default
```
Comment (chrisdoherty4) [inline tinkerbell/stack/values.yaml:32]:
```suggestion
      # The storage type to use for persisting HookOS artifacts. The hostpath
      # option can only be used with single worker node clusters.
      # Options: hostpath, pvc.
      type: hostpath
      
      # The PVC storage class to use. Only valid with type=pvc.
      # storageClassName: default
```

Comment (chrisdoherty4) [reply tinkerbell/stack/values.yaml:32]: I think it may also be useful to include links to Kubernetes documentation.

File: tinkerbell/stack/values.yaml [hunk 6]
```diff
@@ -21,6 +23,24 @@ stack:
...
+
+      ## PVC accessModes : 
+      # Require ReadWriteMany in order to download Hook and scale up kubernetes tink-stack Deployment resource.
```
Comment (chrisdoherty4) [inline tinkerbell/stack/values.yaml:35]: This sentence has little context which makes it a somewhat confusing on what I should set. Specifically, I'm left wondering what ReadWriteMany has to do with scaling a tink-stack deployment and subsequently need to dive into the templates to try and understand.

Perhaps this could be owrded as "If you include ReadWriteMany it will allow X to happen. This will accommodate scaling the Tink Stack because Y".

File: tinkerbell/stack/values.yaml [hunk 7]
```diff
@@ -21,6 +23,24 @@ stack:
...
+      # If ReadWriteOnce is used, you will need to download manually Hook into stack Pod (/usr/share/nginx/html/).
+      accessModes:
+        - ReadWriteMany
+      size: 1Gi
+      # annotations: {}
+      # selectorLabels: {}
```
Comment (chrisdoherty4) [inline tinkerbell/stack/values.yaml:41]:
```suggestion
			# Annotations to include on the PVC object.
      annotations: {}
      
      # An optional selector to narrow the Volumes considered for binding to the PVC.
      # Should be structured the same as a raw PVC selector.
      # See https://kubernetes.io/docs/concepts/storage/persistent-volumes/#selector.
      selector: {}
```

I'm suggesting changing the `selector` so users can define the selector using expressions as well as labels.

File: tinkerbell/stack/values.yaml [hunk 8]
```diff
@@ -21,6 +23,24 @@ stack:
     # downloadURL only works with the > 0.8.1 Hook release because
     # previous Hook versions didn't provide a checksum file.
     downloadURL: https://github.com/tinkerbell/hook/releases/download/latest
+    persistence:
+      enabled: true
+      ## type hostpath or pvc
+      # hostpath : only works on a single worker node cluster
+      type: hostpath
```
Comment (chrisdoherty4) [inline tinkerbell/stack/values.yaml:30]: Sorry to throw a spanner in the works, but when I looked at this again it made me wonder if we'd be better using keys to represent the persistence type avoiding the 'only use this if using pvc' commentary about each of the items below.

E.g.

```yaml
hook:
	persistence:
		hostpath: {}
		pvc:
			storageClassName: ""
			...
```

@jacobweinstock Do you have a preference?

Comment (jacobweinstock) [reply tinkerbell/stack/values.yaml:30]: With your suggestion we could still use `type: ` and then both `hostpath:` and `pvc:` could be populated without issue.

```yaml
hook:
  persistence:
    # type must be either "hostpath" or "pvc"
    # hostpath: only works on a single worker node cluster
    type: hostpath
    hostpath: {}
    pvc:
      storageClassName: ""
      ...
```

Comment (chrisdoherty4) [reply tinkerbell/stack/values.yaml:30]: I'm open to that, though I don't think its strictly necessary because we can test for the presence of either `hostpath` or `pvc` keys. I'm not sure which is clearer so I'm good either way.

**Review by chrisdoherty4 — APPROVED** [review]

File: tinkerbell/stack/templates/_hook.tpl [hunk 9]
```diff
@@ -0,0 +1,14 @@
+{{- define "stack.hookDownloadJobEnabled" -}}
```
Comment (jacobweinstock) [inline tinkerbell/stack/templates/_hook.tpl:1]: can we get a code comment here about what this is and why? Reading the code is a little hard to follow and will require that I re-understand it every time I come back to it. Also, i don't know if I understand it in the first place.

Is this logic to determine if downloading HookOS should occur?
Does this logic say that if there is a pvc with accessMode of ReadWriteOnce then we won't download HookOS? if so, why is that?

File: tinkerbell/stack/templates/hook.yaml [hunk 10]
```diff
@@ -1,4 +1,5 @@
-{{- if .Values.stack.hook.enabled }}
+{{- $hookDownloadJobEnabled := eq (include "stack.hookDownloadJobEnabled" .) "true" -}}
```
Comment (jacobweinstock) [inline tinkerbell/stack/templates/hook.yaml:1]: is this comparison redundant?

true == true is true
false == true is false

couldn't we just use `include "stack.hookDownloadJobEnabled"` directly?

File: tinkerbell/stack/templates/nginx_persistentvolumeclaim.yaml
Comment (jacobweinstock) [inline tinkerbell/stack/templates/nginx_persistentvolumeclaim.yaml:1]: mind renaming this file to `nginx_pvc.yaml`?

File: tinkerbell/stack/templates/NOTES.txt
Comment (jacobweinstock) [inline tinkerbell/stack/templates/NOTES.txt:1]: curious if this empty file provides us something or if it can be removed.

**Review by jacobweinstock — CHANGES_REQUESTED** [review]
@pmorillon, thanks for this. Almost there. I left a few comments for your review. Thank you!

**Comment by jacobweinstock** [comment]
Looks like this branch needs a rebase of main.

File: tinkerbell/stack/templates/_hook.tpl (hunk 9 above)
Comment (pmorillon) [reply tinkerbell/stack/templates/_hook.tpl:1]: > Is this logic to determine if downloading HookOS should occur?

Yes

> Does this logic say that if there is a pvc with accessMode of ReadWriteOnce then we won't download HookOS? if so, why is that?

Because you can't determine on which node will be scheduled the Kubernetes Job, so stack Deployment resource already mounting the PVC, the download job will fail to mount the PVC

File: tinkerbell/stack/templates/hook.yaml (hunk 10 above)
Comment (pmorillon) [reply tinkerbell/stack/templates/hook.yaml:1]: sure

File: tinkerbell/stack/templates/nginx_persistentvolumeclaim.yaml
Comment (pmorillon) [reply tinkerbell/stack/templates/nginx_persistentvolumeclaim.yaml:1]: ok, why not

**Comment by mddeff** [comment]
@chrisdoherty4 @jacobweinstock 

I came to this PR after following #42 and #63 while in search of a solution for hook that didnt use `hostPath`.  It looks like we (and by we, I mean @pmorillon) were SO close to getting this merged.

If the final code-comment you requested is added and re-base complete, can we get this merged? Using `hostPath` could be considered a blocker to running Tinkerbell in many production environments.

I love what the Tinkerbell project is doing, awesome stuff!
